### ISSUE TITLE
3.0.x crash after 256 EAP auths, since update last 2 months

### ISSUE THREAD

**Issue by pauldekkers** [FreeRADIUS/freeradius-server#1927]
# Issue type

- [X] Defect - Crash or memory corruption.

# Defect/Feature description

With a few EAP authentications in a row FreeRADIUS crashes (either a worker or the entire process). The version I got from git in January worked fine, a git clone of 3.0.x today fails!

The crash occurs consistently after 256 auths.

## How to reproduce issue

Clean install on Ubuntu 16.04 LTS, version from "make deb" or really only plain ./configure doesn't matter.
Only changes in users, client.conf, proxy.conf

Tried on two different servers, and compared 3.0.x versions of today and January.

```
# cat users
test	Cleartext-Password := "testing123"

# cat proxy.conf
proxy server {
	default_fallback = no

}
realm test.nl {
}
realm LOCAL {
}

# cat clients.conf
client local {
	ipaddr = 10.0.0.0/24
	secret = testing123
}
```

The rest is default. Then run a series of tests with eapol_test, like:

```
time seq 1 1000 | xargs -n 1 -I{} eapol_test -a10.0.0.159 -p1812 -stesting123 -ctest.conf -N126:s:"1surfnet{}.nl."
```

With test.conf being:

```
network={
        key_mgmt=IEEE8021X
        eap=PEAP
        phase1="peapver=0"
        phase2="auth=MSCHAPv2"
        identity="test@test.nl"
        password="testing123"
        }

```

With the counter Operator-Name I can see the auths fail after about 250 auths with the version of today - while the one in January continues and finishes all auths.

## Output of gdb & ``[radiusd|freeradius] -X`` showing issue occurring

```
(2810) Received Access-Request Id 5 from 10.0.0.160:32901 to 10.0.0.159:1812 length 286
(2810)   User-Name = "test@test.nl"
(2810)   NAS-IP-Address = 127.0.0.1
(2810)   Calling-Station-Id = "02-00-00-00-00-01"
(2810)   Framed-MTU = 1400
(2810)   NAS-Port-Type = Wireless-802.11
(2810)   Connect-Info = "CONNECT 11Mbps 802.11b"
(2810)   Operator-Name = "1surfnet256.nl."
(2810)   EAP-Message = 0x0205008819800000007e16030300461000004241045538f0e9b6bf2d58d6ef9776fc73a80ce7ec32e22c07badf12efa0204aa0f594b35007eabe6aef4b8adde6ea270c8031afeb0e4f74133df57e8f8b29b1b2838d1403030001011603030028adb19243a38d8e817f5e75f210af0abf9317cf8b6ee1a9
(2810)   State = 0x89ed34e08de82d57ca5e07425c8f987d
(2810)   Message-Authenticator = 0xae1f9489f4d45638ebea62b164866d7d
(2810) session-state: No cached attributes
(2810) # Executing section authorize from file /opt/frs3/etc/raddb/sites-enabled/default
(2810)   authorize {
(2810)     policy filter_username {
(2810)       if (&User-Name) {
(2810)       if (&User-Name)  -> TRUE
(2810)       if (&User-Name)  {
(2810)         if (&User-Name =~ / /) {
(2810)         if (&User-Name =~ / /)  -> FALSE
(2810)         if (&User-Name =~ /@[^@]*@/ ) {
(2810)         if (&User-Name =~ /@[^@]*@/ )  -> FALSE
(2810)         if (&User-Name =~ /\.\./ ) {
(2810)         if (&User-Name =~ /\.\./ )  -> FALSE
(2810)         if ((&User-Name =~ /@/) && (&User-Name !~ /@(.+)\.(.+)$/))  {
(2810)         if ((&User-Name =~ /@/) && (&User-Name !~ /@(.+)\.(.+)$/))   -> FALSE
(2810)         if (&User-Name =~ /\.$/)  {
(2810)         if (&User-Name =~ /\.$/)   -> FALSE
(2810)         if (&User-Name =~ /@\./)  {
(2810)         if (&User-Name =~ /@\./)   -> FALSE
(2810)       } # if (&User-Name)  = notfound
(2810)     } # policy filter_username = notfound
(2810)     [preprocess] = ok
(2810)     [chap] = noop
(2810)     [mschap] = noop
(2810)     [digest] = noop
(2810) suffix: Checking for suffix after "@"
(2810) suffix: Looking up realm "test.nl" for User-Name = "test@test.nl"
(2810) suffix: Found realm "test.nl"
(2810) suffix: Adding Stripped-User-Name = "test"
(2810) suffix: Adding Realm = "test.nl"
(2810) suffix: Authentication realm is LOCAL
(2810)     [suffix] = ok
(2810) eap: Peer sent EAP Response (code 2) ID 5 length 136
(2810) eap: Continuing tunnel setup
(2810)     [eap] = ok
(2810)   } # authorize = ok
(2810) Found Auth-Type = eap
(2810) # Executing group from file /opt/frs3/etc/raddb/sites-enabled/default
(2810)   authenticate {
(2810) eap: Expiring EAP session with state 0x89ed34e08de82d57
(2810) eap: Finished EAP session with state 0x89ed34e08de82d57
(2810) eap: Previous EAP request found for state 0x89ed34e08de82d57, released from the list
(2810) eap: Peer sent packet with method EAP PEAP (25)
(2810) eap: Calling submodule eap_peap to process data
(2810) eap_peap: Continuing EAP-TLS
(2810) eap_peap: Peer indicated complete TLS record size will be 126 bytes
(2810) eap_peap: Got complete TLS record (126 bytes)
(2810) eap_peap: [eaptls verify] = length included
(2810) eap_peap: <<< recv TLS 1.2  [length 0046]
(2810) eap_peap: TLS_accept: unknown state
(2810) eap_peap: TLS_accept: unknown state
(2810) eap_peap: <<< recv TLS 1.2  [length 0001]
(2810) eap_peap: <<< recv TLS 1.2  [length 0010]
(2810) eap_peap: TLS_accept: unknown state
(2810) eap_peap: >>> send TLS 1.2  [length 0001]
(2810) eap_peap: TLS_accept: unknown state
(2810) eap_peap: >>> send TLS 1.2  [length 0010]
(2810) eap_peap: TLS_accept: unknown state
(2810) eap_peap: TLS_accept: unknown state
tls: Freeing cached session VPs
src/lib/pair.c:2286: Type mismatch: name[rbnode_t] expected[VALUE_PAIR]

talloc abort: src/lib/pair.c:2286: Type mismatch: name[rbnode_t] expected[VALUE_PAIR]

Backtrace of last 40 frames:
/opt/frs3/lib/libfreeradius-radius.so(+0xd5b7)[0x7ffff79555b7]
/usr/lib/x86_64-linux-gnu/libtalloc.so.2(_talloc_get_type_abort+0xb7)[0x7ffff708fee7]
/opt/frs3/lib/libfreeradius-radius.so(fr_pair_verify+0x85)[0x7ffff7967a39]
/opt/frs3/lib/libfreeradius-radius.so(fr_cursor_init+0x7e)[0x7ffff7953d67]
/opt/frs3/lib/libfreeradius-radius.so(fr_pair_list_free+0x41)[0x7ffff7963efe]
/opt/frs3/sbin/radiusd[0x4592cc]
/lib/x86_64-linux-gnu/libcrypto.so.1.0.0(+0x649f6)[0x7ffff75689f6]
/lib/x86_64-linux-gnu/libssl.so.1.0.0(SSL_SESSION_free+0x4c)[0x7ffff72ddbcc]
/lib/x86_64-linux-gnu/libssl.so.1.0.0(+0x43532)[0x7ffff72de532]
/lib/x86_64-linux-gnu/libssl.so.1.0.0(SSL_CTX_add_session+0x141)[0x7ffff72de6f1]
/lib/x86_64-linux-gnu/libssl.so.1.0.0(+0x3ef60)[0x7ffff72d9f60]
/lib/x86_64-linux-gnu/libssl.so.1.0.0(+0x1b345)[0x7ffff72b6345]
/lib/x86_64-linux-gnu/libssl.so.1.0.0(+0x27758)[0x7ffff72c2758]
/lib/x86_64-linux-gnu/libssl.so.1.0.0(+0x23c64)[0x7ffff72bec64]
/opt/frs3/sbin/radiusd(tls_handshake_recv+0x113)[0x454b6a]
/opt/frs3/lib/libfreeradius-eap.so(+0x7053)[0x7ffff53f3053]
/opt/frs3/lib/libfreeradius-eap.so(eaptls_process+0x44a)[0x7ffff53f3554]
/opt/frs3/lib/rlm_eap_peap.so(+0x1fdd)[0x7ffff1533fdd]
/opt/frs3/lib/rlm_eap.so(+0x41ac)[0x7ffff55fe1ac]
/opt/frs3/lib/rlm_eap.so(eap_method_select+0x430)[0x7ffff55feb8e]
/opt/frs3/lib/rlm_eap.so(+0x2d59)[0x7ffff55fcd59]
/opt/frs3/sbin/radiusd[0x42afd7]
/opt/frs3/sbin/radiusd[0x42b728]
/opt/frs3/sbin/radiusd[0x42b19f]
/opt/frs3/sbin/radiusd[0x42c217]
/opt/frs3/sbin/radiusd(modcall+0xb1)[0x42d089]
/opt/frs3/sbin/radiusd(indexed_modcall+0x366)[0x4284ea]
/opt/frs3/sbin/radiusd(process_authenticate+0x22)[0x42a8c7]
/opt/frs3/sbin/radiusd[0x40f72f]
/opt/frs3/sbin/radiusd(rad_authenticate+0x59e)[0x41014d]
/opt/frs3/sbin/radiusd[0x440756]
/opt/frs3/sbin/radiusd[0x43f449]
/opt/frs3/sbin/radiusd(request_receive+0x7ee)[0x440fe8]
/opt/frs3/sbin/radiusd[0x41942f]
/opt/frs3/sbin/radiusd[0x448716]
/opt/frs3/lib/libfreeradius-radius.so(fr_event_loop+0x651)[0x7ffff797fb40]
/opt/frs3/sbin/radiusd(radius_event_process+0x26)[0x44a6db]
/opt/frs3/sbin/radiusd(main+0xc90)[0x433bb3]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf0)[0x7ffff6685830]
/opt/frs3/sbin/radiusd(_start+0x29)[0x40eb69]

Program received signal SIGABRT, Aborted.
0x00007ffff669a428 in __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:54
54	../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) bt
#0  0x00007ffff669a428 in __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:54
#1  0x00007ffff669c02a in __GI_abort () at abort.c:89
#2  0x00007ffff7955600 in _fr_talloc_fault_simple (reason=0xd32120 "src/lib/pair.c:2286: Type mismatch: name[rbnode_t] expected[VALUE_PAIR]") at src/lib/debug.c:782
#3  0x00007ffff708fee7 in _talloc_get_type_abort () from /usr/lib/x86_64-linux-gnu/libtalloc.so.2
#4  0x00007ffff7967a39 in fr_pair_verify (file=0x7ffff7980fd0 "src/lib/cursor.c", line=77, vp=0x992060) at src/lib/pair.c:2286
#5  0x00007ffff7953d67 in fr_cursor_init (cursor=0x7fffffffc210, const_vp=0x7fffffffc280) at src/lib/cursor.c:77
#6  0x00007ffff7963efe in fr_pair_list_free (vps=0x7fffffffc280) at src/lib/pair.c:159
#7  0x00000000004592cc in sess_free_vps (parent=0x990680, data_ptr=0x992060, ad=0x990778, idx=1, argl=0, argp=0x0) at src/main/tls.c:2456
#8  0x00007ffff75689f6 in ?? () from /lib/x86_64-linux-gnu/libcrypto.so.1.0.0
#9  0x00007ffff72ddbcc in SSL_SESSION_free () from /lib/x86_64-linux-gnu/libssl.so.1.0.0
#10 0x00007ffff72de532 in ?? () from /lib/x86_64-linux-gnu/libssl.so.1.0.0
#11 0x00007ffff72de6f1 in SSL_CTX_add_session () from /lib/x86_64-linux-gnu/libssl.so.1.0.0
#12 0x00007ffff72d9f60 in ?? () from /lib/x86_64-linux-gnu/libssl.so.1.0.0
#13 0x00007ffff72b6345 in ?? () from /lib/x86_64-linux-gnu/libssl.so.1.0.0
#14 0x00007ffff72c2758 in ?? () from /lib/x86_64-linux-gnu/libssl.so.1.0.0
#15 0x00007ffff72bec64 in ?? () from /lib/x86_64-linux-gnu/libssl.so.1.0.0
#16 0x0000000000454b6a in tls_handshake_recv (request=0xa77ef0, ssn=0x96d9d0) at src/main/tls.c:709
#17 0x00007ffff53f3053 in eaptls_operation (status=FR_TLS_LENGTH_INCLUDED, handler=0xb69c10) at src/modules/rlm_eap/libeap/eap_tls.c:705
#18 0x00007ffff53f3554 in eaptls_process (handler=0xb69c10) at src/modules/rlm_eap/libeap/eap_tls.c:901
#19 0x00007ffff1533fdd in mod_process (arg=0x921850, handler=0xb69c10) at src/modules/rlm_eap/types/rlm_eap_peap/rlm_eap_peap.c:265
#20 0x00007ffff55fe1ac in eap_module_call (module=0x919bb0, handler=0xb69c10) at src/modules/rlm_eap/eap.c:194
#21 0x00007ffff55feb8e in eap_method_select (inst=0x8ee830, handler=0xb69c10) at src/modules/rlm_eap/eap.c:457
#22 0x00007ffff55fcd59 in mod_authenticate (instance=0x8ee830, request=0xa77ef0) at src/modules/rlm_eap/rlm_eap.c:288
#23 0x000000000042afd7 in call_modsingle (component=MOD_AUTHENTICATE, sp=0x942e20, request=0xa77ef0) at src/main/modcall.c:302
#24 0x000000000042b728 in modcall_recurse (request=0xa77ef0, component=MOD_AUTHENTICATE, depth=1, entry=0x7fffffffd778, do_next_sibling=true) at src/main/modcall.c:578
#25 0x000000000042b19f in modcall_child (request=0xa77ef0, component=MOD_AUTHENTICATE, depth=1, entry=0x7fffffffd760, c=0x942e20, result=0x7fffffffd0f0, do_next_sibling=true) at src/main/modcall.c:408
#26 0x000000000042c217 in modcall_recurse (request=0xa77ef0, component=MOD_AUTHENTICATE, depth=0, entry=0x7fffffffd760, do_next_sibling=true) at src/main/modcall.c:789
#27 0x000000000042d089 in modcall (component=MOD_AUTHENTICATE, c=0x942d20, request=0xa77ef0) at src/main/modcall.c:1134
#28 0x00000000004284ea in indexed_modcall (comp=MOD_AUTHENTICATE, idx=12259193, request=0xa77ef0) at src/main/modules.c:1028
#29 0x000000000042a8c7 in process_authenticate (auth_type=12259193, request=0xa77ef0) at src/main/modules.c:2168
#30 0x000000000040f72f in rad_check_password (request=0xa77ef0) at src/main/auth.c:252
#31 0x000000000041014d in rad_authenticate (request=0xa77ef0) at src/main/auth.c:571
#32 0x0000000000440756 in request_running (request=0xa77ef0, action=1) at src/main/process.c:1535
#33 0x000000000043f449 in request_queue_or_run (request=0xa77ef0, process=0x4405f4 <request_running>) at src/main/process.c:1023
#34 0x0000000000440fe8 in request_receive (ctx=0xa77c10, listener=0x95bca0, packet=0xa77c70, client=0x8510f0, fun=0x40fbaf <rad_authenticate>) at src/main/process.c:1791
#35 0x000000000041942f in auth_socket_recv (listener=0x95bca0) at src/main/listen.c:1571
#36 0x0000000000448716 in event_socket_handler (xel=0x852ef0, fd=8, ctx=0x95bca0) at src/main/process.c:4593
#37 0x00007ffff797fb40 in fr_event_loop (el=0x852ef0) at src/lib/event.c:649
#38 0x000000000044a6db in radius_event_process () at src/main/process.c:5666
#39 0x0000000000433bb3 in main (argc=2, argv=0x7fffffffe698) at src/main/radiusd.c:585
(gdb)
```

**Comment by pauldekkers** [comment]
Going back to 9796921a3c32de63c9ea63786fdb2694c91b25d0 resolves it.

**Comment by alandekok** [comment]
I've pushed a fix to the v3.0.x head.  Please try that.

**Comment by pauldekkers** [comment]
Thanks, that resolves it.
